### ISSUE TITLE
[UX] Block UI: Dynamically load second-step visibility conditions configuration in the same dialog.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1020]
As it is now, a user selects to add a condition:

![block_config-add_condition](https://cloud.githubusercontent.com/assets/2423362/8270538/afc8ea1e-1825-11e5-9dc9-ca5589089287.png)

...and that starts another dialog in order to select from the various available condition categories:

![block_config-conditions](https://cloud.githubusercontent.com/assets/2423362/8270535/90196e5a-1825-11e5-8740-56d6abd7407b.png)

...and then yet another dialog to configure any options for the selected condition:

![block_config-configure_condition](https://cloud.githubusercontent.com/assets/2423362/8270540/d9705c6c-1825-11e5-8f1c-8b54dd98b075.png)

There's a few things that feel wrong here:

- "Jumping" to a new dialogue after each step (after each form submit).
- Both the 2nd and 3rd step have "Add condition" as primary button. I had the feeling that hitting it the first time would be it, but there was another step involved.
- We already have a similar way (using a modal) in the Views UI that lets the user configure things, but that feels more like a one-step thing. It actually is not (because when you select an option, then further options appear), but the fact that you hit a submit button only once gives you that sense. I think this is better.

So, can we implement the same UI we have in Views here as well?

**Comment by quicksketch** [comment]
> We already have a similar way (using a modal) in the Views UI that lets the user configure things, but that feels more like a one-step thing. It actually is not (because when you select an option, then further options appear), but the fact that you hit a submit button only once gives you that sense. I think this is better.

Layouts are fairly similar to Views in that the first dialog asks which field/filter/etc should be added, then the second dialog asks for configuration. It just so happens that you're already in a dialog when adding the condition, so we end up with 3 dialogs instead of 2.

But not to belabor the point, it would indeed be preferable if we could eliminate the 3rd dialog and merge the options into the 2nd one where you select the conditional. We could probably load the form options via AJAX below the conditional select list. This actually makes it *less* like Views, but I don't think consistency should be overriding our attempts to make the best UI in each situation as possible.

**Comment by klonos** [comment]
Fair enough.

**Comment by docwilmot** [comment]
Fixing #943, almost there, but realised this issue conflicts with that one a bit. Thought of a question though: why do we need the "Add visibility condition" link at all? All it does is popup a form with a select. Why not put the select right there? Is this a viable option? 

Now:
![visib now](https://cloud.githubusercontent.com/assets/1042283/12868659/d2c216b8-ccd8-11e5-94db-6a0ba7f0f8a3.PNG)

Instead:
![visib select](https://cloud.githubusercontent.com/assets/1042283/12868646/9e746e38-ccd8-11e5-800e-6ae5cbf6d36e.PNG)

@klonos? Anyone?

**Comment by docwilmot** [comment]
Also, for fixing the main point of this issue, I think actually the UX of loading the last step in its own dialog actually is best. It doesnt seem quite workable otherwise;; in @klonos 2nd image above for example, if we were to load the third image into there, we'd be left with the select element plus the fields to configure the one you just selected. It just simply is cleaner to load the "Configure condition" form in a 2nd dialog IMHO.

To mitigate the issue @klonos was having, we could, for the 2nd dialog, get rid of the "Add condition" button, and just load options once the user clicks the option in the select. Or change the select to a  plain list of links and load on click. So you only reach the "Add condition" button in the last step, which makes more sense.

So. in conclusion:
1. Leave everything as is; or
2. Load the "Configure condition" dialog into the "Add condition for XXX" dialog as per this issue; OR
3. Leave the dialog train as it is, but remove the "Add condition" button from the "Configure condition for XX" dialog.

I'm for 1 or 3 actually

**Comment by docwilmot** [comment]
@klonos @jenlampton @quicksketch @mikemccaffrey I'm on a mission to finish the layout followups; this one is blocking my last PR. Can I get a decision please?

**Comment by klonos** [comment]
@docwilmot are you suggesting loosing the dialog entirely and using #states to show the additional options that go with each drop-down menu item instead? If so, I'm not sure how well that would that work with multi-step dialogs. Also, what about adding multiple conditions? Would you show an additional drop-down after the user made a selection in the first?

> ...we could, for the 2nd dialog, get rid of the "Add condition" button, and just load options once the user clicks the option in the select.

Thought of that, but what if the user made an accidental click? A set of "next"/"back" buttons? Hmm not good UX IMO. While if the options are loaded in the same page, the drop-down is still there available for the user to change the initial selection.

> Or change the select to a plain list of links and load on click. So you only reach the "Add condition" button in the last step, which makes more sense.

I generally like the solution of select lists or radio buttons over drop-downs because they take a single click instead of two. In this case though, I'd have to insist that the loading of the additional options should happen on the same page in order to give the user the option to change their mind if they just made an accidental click. Plus, if all the options are in a single dialog (in this case of course, where we don't have a great deal of them), it gives an overview before the user finally clicks on the "Add condition" button.

PS: sorry for the late reply. Browser crash with many pending tabs open.

**Comment by docwilmot** [comment]
> I'd have to insist that the loading of the additional options should happen on the same page in order to give the user the option to change their mind if they just made an accidental click

Why insist? They could just press the cancel button, assuming that the cancel button on the final form brings you back to the select rather than just closing the dialog, which will be fixed as well. 

I really can't see a good way to open the options in the same dialog. Would this open below the select list? Above? If the list is long then the form would be below the fold, no?

**Comment by klonos** [comment]
> ...assuming that the cancel button on the final form brings you back to the select rather than just closing the dialog,

Precisely because of that. The cancel button is not a "back" step button. Having it behave as such would break UI patterns that users are used to. That's why I mentioned in my comment:

> ...A set of "next"/"back" buttons?...

...in order to have this resemble a wizard thingy. But not such good UX as I said.

> I really can't see a good way to open the options in the same dialog.

AJAX or #states??

> ...Would this open below the select list? Above?

I'd say below is again a pattern people are used to when it comes to more options/content loading on a page or dialog.

> If the list is long then the form would be below the fold, no?

Yes, but that could happen anyways with small (short) viewports. Besides we already have cases where the dialog height is not enough and what we do then is add scrollbars. The "Add block" dialog for example with the long list of available blocks:

![backdrop-example_of_dialog_with_scrollbars-add_block](https://cloud.githubusercontent.com/assets/2423362/13066431/ec871bc8-d4b9-11e5-9d31-56d49187c2be.png)

It won't be that bad I think:

![backdrop-visibility_conditions_dialog-merged_steps-front_page](https://cloud.githubusercontent.com/assets/2423362/13066613/1526edfe-d4bc-11e5-84e2-21bcc01aae8e.png)
![backdrop-visibility_conditions_dialog-merged_steps-language](https://cloud.githubusercontent.com/assets/2423362/13066617/154a33fe-d4bc-11e5-8a14-0d69e3142a3a.png)
![backdrop-visibility_conditions_dialog-merged_steps-path](https://cloud.githubusercontent.com/assets/2423362/13066614/152ec9ac-d4bc-11e5-9cd4-3709e5d7c3c7.png)
![backdrop-visibility_conditions_dialog-merged_steps-permission](https://cloud.githubusercontent.com/assets/2423362/13066615/15393202-d4bc-11e5-8951-16dc74eeb18e.png)
![backdrop-visibility_conditions_dialog-merged_steps-role](https://cloud.githubusercontent.com/assets/2423362/13066616/153d3230-d4bc-11e5-939a-9f9400d44a30.png)

**Comment by docwilmot** [comment]
Touche. Images help; makes sense now.

**Comment by klonos** [comment]
> I'm on a mission to finish the layout followups; this one is blocking my last PR.

Glad I helped then :wink:

**Comment by quicksketch** [comment]
Great work @docwilmot! This change is most welcome! I tested out the new PR at https://github.com/backdrop/backdrop/issues/1260. Adding a new condition is *much* faster with the new AJAX-enabled form.

The only issue I've found right away is that the "Configure" link for editing existing conditions doesn't seem to work any more. When clicking, nothing happens in the UI. In the network inspector, I can see that the AJAX request is encountering a fatal error: `Fatal error: Class name must be a valid object or a string in /home/nate/Sites/backdrop/core/modules/layout/layout.module on line 1294`

This is looking really good!

**Comment by quicksketch** [comment]
As @docwilmot pointed out in the PR, this change effectively fixes https://github.com/backdrop/backdrop-issues/issues/943 and https://github.com/backdrop/backdrop-issues/issues/1022. I've closed those issues as duplicates of this one.

**Comment by quicksketch** [comment]
Thanks for the updated PR @docwilmot! The dialogs now work properly when configuring an existing condition.

I encountered a small oddity in that if I configure an existing condition, it still says "Add visibility condition" in the dialog title, help, and the submit button. Although it's minor from a technical stand-point, it would really help from a UX perspective. I had a hard time remembering if I was in the middle of configuring an existing condition or adding a brand new one.

I found a new bug as well:

- Add a new condition to a block/menu item/layout. Configure the condition and close the dialog.
- Configure the newly added condition. Change it's type to a different one.
- Instead of replacing the existing condition, a second condition is added. The first condition is left unaltered.

I've tested some fixes and I'll file a PR against your current branch @docwilmot.

**Comment by quicksketch** [comment]
@docwilmot If you can review https://github.com/docwilmot/backdrop/pull/4, that should fix the issues I found. Please feel free to adjust if needed.

**Comment by quicksketch** [comment]
@docwilmot has fixed this in https://github.com/backdrop/backdrop/issues/1260. Great job! This knocks out three issues with one fix! This will be in 1.4.0.

**Comment by quicksketch** [comment]
This PR seems to have caused our tests to fail after merging other PRs since this one was filed. I'll make a quick PR to fix this test regression.

**Comment by quicksketch** [comment]
The failure we're getting now:

```
---- Layout: Layout Interface Test (LayoutInterfaceTest) ----
Status    Group      Filename          Line Function                            
--------------------------------------------------------------------------------
Fail      Other      layout.test        489 LayoutInterfaceTest->testBlockEntit
    Failed to set field entity_id to 1
```

**Comment by quicksketch** [comment]
Failing tests fixed in https://github.com/backdrop/backdrop/pull/1376.

**Comment by jenlampton** [comment]
@docwilmot++ you are amazing. 👍
